src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+
+            // Do a pass to eliminate blocks without statements and only Next set.
+            Pack(blocks);
+
+            Debug.Assert(builder._evalStack.Count == 0);
+            builder._evalStack.Free();
+
+            return blocks.ToImmutableAndFree();
+        }
+
+        private static void Pack(ArrayBuilder<BasicBlock> blocks)
+        {
+            int count = blocks.Count - 1;
+            for (int i = 1; i < count; i++)
+            {
+                BasicBlock block = blocks[i];
+                Debug.Assert(block.Next != null);
+                if (block.Statements.IsEmpty && block.Conditional.Condition == null)
+                {
+                    BasicBlock next = block.Next;
+                    foreach (BasicBlock predecessor in block.Predecessors)
+                    {
+                        if (predecessor.Next == block)
+                        {
+                            predecessor.Next = next;
+                            next.AddPredecessor(predecessor);
+                        }
+
+                        (IOperation condition, bool jumpIfTrue, BasicBlock destination) = predecessor.Conditional;
+                        if (destination == block)
+                        {
+                            predecessor.Conditional = (condition, jumpIfTrue, next);
+                            next.AddPredecessor(predecessor);
+                        }
+                    }
+
+                    next.RemovePredecessor(block);
+                    blocks.RemoveAt(i);
+                    i--;
+                    count--;
+                }
+            }
+        }
+
+        private void VisitStatement(IOperation operation)
+        {
+            IOperation saveCurrentStatement = _currentStatement;
+            _currentStatement = operation;
+            Debug.Assert(_evalStack.Count == 0);
+
+            // PROTOTYPE(dataflow): Ensure that statement's parent is null at this point.
+            AddStatement(Visit(operation, null));
+            Debug.Assert(_evalStack.Count == 0);
+            _currentStatement = saveCurrentStatement;
+        }
+
+        private BasicBlock CurrentBasicBlock
+        {
+            get
+            {
+                if (_currentBasicBlock == null)
+                {
+                    AppendNewBlock(new BasicBlock(BasicBlockKind.Block));
+                }
+
+                return _currentBasicBlock;
+            }
+        }
+
+        private void AddStatement(IOperation statement)
+        {
+            if (statement == null)
+            {
+                return;
+            }
+
+            // PROTOTYPE(dataflow): Assert that statement's parent is null at this point.
+            CurrentBasicBlock.AddStatement(statement);
+        }
+
+        private void AppendNewBlock(BasicBlock block)
+        {
+            Debug.Assert(block != null);
+            BasicBlock prevBlock = _blocks.Last();
+
+            if (prevBlock.Next == null)
```

Comment:
Is it possible for `prevBlock.Next` to be non-null here?

Reply:
> Is it possible for prevBlock.Next to be non-null here?

Yes, for example a goto statement will set the Next link. 

---
In reply to: [162151898](https://github.com/dotnet/roslyn/pull/24263#discussion_r162151898) [](ancestors = 162151898)